DNN Platform/DotNetNuke.Web.Mvc/Framework/Controllers/IDnnController.cs

```diff
@@ -28,5 +29,7 @@ public interface IDnnController : IController
 
         ViewEngineCollection ViewEngineCollectionEx { get; set; }
 
+        DnnUrlHelper Url { get; set; }
```

Comment:
Why do we need this property here? System.Web.Mvc.Controller already has the Url property which is being overridden by using the new keyword in DnnController.

Reply:
As far as I know the `DnnUrlHelper` does not inherit from the `System.Web.Mvc.UrlHelper` `UrlHelper` which are different objects. This is why I chose to create the `DnnUrlHelper` inside the `IDnnController`. This way if someone is using the interface they will have access to the appropriate `DnnUrlHelper`. 

See [DnnUrlHelper](https://github.com/dnnsoftware/Dnn.Platform/blob/development/DNN%20Platform/DotNetNuke.Web.Mvc/Helpers/DnnUrlHelper.cs) from the platform code. Snippet below:

```c#
namespace DotNetNuke.Web.Mvc.Helpers
{
    public class DnnUrlHelper
    {
```

Taking a closer look at the `System.Web.Mvc` [UrlHelper](https://github.com/aspnet/AspNetWebStack/blob/62d0b2df0bcca848a8e0848fd1866928df15528e/src/System.Web.Mvc/UrlHelper.cs). It appears we can take our current `DnnUrlHelper` and inherit from it and override all the `Action` methods. Without making this change we will not be able to properly override the `Url` property in the `DnnController`

Here are some proposed changes based off the feedback provided:

* Update the `DnnUrlHelper` to inherit and override the `System.Web.Mvc.UrlHelper`
* Remove the `DnnUrlHelper` from the `IDnnController`
  * This can be replaced with the base class `System.Web.Mvc.UrlHelper` or omitted completly
* We should be able to still override the `Initialize()` method and set the `System.Web.Mvc.UrlHelper` equal to our implementation of `DnnUrlHelper`

If these changes all make sense to you, I'll go ahead and update the PR

Reply:
@ahoefling Regarding `DnnUrlHelper` in `IDnnController `since you are using new and same property name, it automatically hides the `UrlHelper `from `System.Web.Mvc.Controller`. We are doing same for `User `property as well. Further, I can agree on the point of reuse the helper if `IDnnController `is used but I don't see much usage of implementing it alone. But we can still leave it the way you did it assuming it to be better way of hiding the property in this scenario.

Regarding inheriting from `UrlHelper`, We can do that way but we can not just override the implementation of `Action `methods and leave others untouched. If we want to inherit from `UrlHelper `then all other methods should be overridden and implemented in a way that they are useful for DNN MVC framework. We don't want to include methods in `DnnUrlHelper `which are not useful and will create incorrect results. 
If you still want to use that approach then that will need to implement all the methods from `UrlHelper`, verify the proper working for them and add test cases for all of them.

Reply:
@amarjit-dhunna  wrote:
>Regarding `DnnUrlHelper` in `IDnnControllersince` you are using new and same property name, it automatically hides the `UrlHelper` from `System.Web.Mvc.Controller`. We are doing same for `User` property as well.

It sounds like the first approach is an established best practice that we have used for the `User` object. Based on your feedback I think it will be best to use the new opperator to hide the the base implementation of `UrlHelper` so module devs will get just the working `DnnUrlHelper`.

I believe it to be best to keep the `IDnnController` the way currently proposed in this PR. Since the `Controller` class uses a different object for the `Url` property (`UrlHelper` vs `DnnUrlHelper`) it will be difficult to write any generic code around the `IDnnController` where the developer would want to leverage the `DnnUrlHelper`. 

@amarjit-dhunna  wrote:
>Regarding inheriting from `UrlHelper`, We can do that way but we can not just override the implementation of `Action` methods and leave others untouched.

I completely agree with you and while this may be a great solution in the future i think it would be best handled in another PR that expands upon the changes here. This may be something I am willing to pick up after we get through this change.

Action Items I have gathered from this discussion:

* Lets keep the PR as proposed including leaving the reference to `IDnnUrlHelper` on the `IDnnController`
* Add unit tests as necessary to cover new code paths

Please sign off on this direction if you are okay with it and I'll update the PR when ready

Reply:
Yes. So only task pending for you is to add unit tests to cover the new code.